src/schema.c

```diff
@@ -113,14 +123,93 @@ ldap_schema_object_classes(const ldap_schema_t *schema)
 LDAPAttributeType**
 ldap_schema_attribute_types(const ldap_schema_t* schema)
 {
-    if (!schema)
-    {
-        ld_error("Schema is NULL.\n");
+    check_for_null_parameter(schema, "ldap_schema_attribute_types - schema is NULL!\n");
+    check_for_null_parameter(schema->attribute_types_by_oid,
+                             "ldap_schema_attribute_types - attribute_types_by_oid is NULL!\n");
 
-        return NULL;
+    int result_size = g_hash_table_size(schema->attribute_types_by_oid);
+
+    LDAPAttributeType** result = talloc_array(schema, LDAPAttributeType*, result_size + 1);
+
+    check_for_null_parameter(result,
+                             "ldap_schema_attribute_types - talloc_array for LDAPAttributeType returned NULL!\n");
+
+    GHashTableIter iter;
+    gpointer key = NULL, value = NULL;
+
+    int index = 0;
+    g_hash_table_iter_init(&iter, schema->attribute_types_by_oid);
+    while (g_hash_table_iter_next(&iter, &key, &value)) {
+        result[index] = value;
+        index++;
     }
+    result[result_size] = NULL;
+
+    return result;
+}
+
+/*!
+ * \brief ldap_schema_get_objectclass_by_oid    Returns a LDAPObjectClass struct by oid key.
+ * \param[in] schema                            Schema to work with.
+ * \return
+ *        - NULL if schema is NULL.
+ *        - Object class from schema.
+ */
+LDAPObjectClass *ldap_schema_get_objectclass_by_oid(const ldap_schema_t* schema, const char *oid)
+{
+    check_for_null_parameter(schema, "ldap_schema_get_objectclass_by_oid - schema is NULL!\n");
+    check_for_null_parameter(schema->object_classes_by_oid,
+                             "ldap_schema_get_objectclass_by_oid - object_classes_by_oid is NULL!\n");
+
+    return (LDAPObjectClass *)g_hash_table_lookup(schema->object_classes_by_oid, oid);
+}
+
+/*!
+ * \brief ldap_schema_get_objectclass_by_name   Returns a LDAPObjectClass struct by name key.
+ * \param[in] schema                            Schema to work with.
+ * \return
+ *        - NULL if schema is NULL.
+ *        - Object class from schema.
+ */
+LDAPObjectClass *ldap_schema_get_objectclass_by_name(const ldap_schema_t* schema, const char *name)
+{
+    check_for_null_parameter(schema, "ldap_schema_get_objectclass_by_name - schema is NULL!\n");
+    check_for_null_parameter(schema->object_classes_by_name,
+                             "ldap_schema_get_objectclass_by_name - object_classes_by_name is NULL!\n");
+
+    return (LDAPObjectClass *)g_hash_table_lookup(schema->object_classes_by_name, name);
+}
+
+/*!
+* \brief ldap_schema_get_attributetype_by_oid  Returns a LDAPAttributeType struct by oid key.
```

Comment:
Replace description with `Returns a LDAPAttributeType struct corresponding to the specified OID.`